R/writeModel.R

```diff
@@ -25,62 +23,71 @@
-##' @return Nothing, but as a side effect, the model file is written
 ##'
 ##' @export
-##' @author original idea by Jouni Kerman, modified by Uwe Ligges, DSB removed S
-##' Plus part
-writeModel <- function(model, con = "model.bug", digits = 5)
-{
-    if (is.R()){
-        model.text <- c("model", replaceScientificNotationR(body(model),
-                                                            digits = digits))
-                                        # "[\+\-]?\d*\.?[Ee]?[\+\-]?\d*"
-    } else {
-        ## In S-PLUS the source code of a function can be obtained with
-        ## as.character(function_name).
-        ## This omits the "function_name <- function()" piece
-        model.text <- paste("model", as.character(model))
-    }
-    model.text <- gsub("%_%", "", model.text)
-    writeLines(model.text, con = con)
+##' @author original idea by Jouni Kerman, modified by Uwe Ligges,
+##'  DSB removed S-Plus part, WW modified the implementation.
+writeModel <- function(model, con = "model.bug", digits = 5) { # nolintr
+  if (is.R()) {
+    model_text <- c("model", h_replace_scientific(model, digits = digits))
+  } else {
+    # In S-PLUS the body of a function can be obtained with
+    # as.character(function_name).
+    model_text <- paste("model", as.character(model))
+  }
+  model_text <- gsub("%_%", "", model_text)
+  writeLines(model_text, con = con)
 }
 
-##' @keywords internal
-replaceScientificNotationR <- function(bmodel, digits = 5){
-    env <- new.env()
-    assign("rSNRidCounter", 0, envir=env)
-    replaceID <- function(bmodel, env, digits = 5){
-        for(i in seq_along(bmodel)){
-            if(length(bmodel[[i]]) == 1){
-                if(as.character(bmodel[[i]]) %in% c(":", "[", "[[")) return(bmodel)
-                if((typeof(bmodel[[i]]) %in% c("double", "integer")) && ((abs(bmodel[[i]]) < 1e-3) || (abs(bmodel[[i]]) > 1e+4))){
-                    counter <- get("rSNRidCounter", envir=env) + 1
-                    assign("rSNRidCounter", counter, envir=env)
-                    id <- paste("rSNRid", counter, sep="")
-                    assign(id, formatC(bmodel[[i]], digits=digits, format="E"), envir=env)
-                    bmodel[[i]] <- id
-                }
-            } else {
-                bmodel[[i]] <- replaceID(bmodel[[i]], env, digits = digits)
-            }
-        }
-        bmodel
-    }
-    bmodel <- deparse(replaceID(bmodel, env, digits = digits), control = NULL)
-    for(i in ls(env)){
-        bmodel <- gsub(paste('"', i, '"', sep=''), get(i, envir=env), bmodel, fixed=TRUE)
+h_replace_scientific <- function(model, digits = 5) {
+  assert_function(model)
+  assert_int(digits)
+
+  bmodel <- h_bapply(
+    body(model),
+    h_format_number,
+    classes = c("integer", "numeric"),
+    digits
+  )
+  bmodel_text <- deparse(bmodel, control = NULL)
+  gsub("\"TEMP_NUM_PREF_|_TEMP_NUM_SUF\"", "", bmodel_text)
+}
+
+# Body apply
+h_bapply <- function(b, f, classes, ...) {
```

Comment:
try to use longer and speaking argument names, e.g. `body`, `fun`

Reply:
also assert on the `body` argument - what should it be?

Reply:
Thanks Daniel! Actually I had a small 2 problems wit this: 
1. If I use `body` as argument name, I will cover the `body` function from `base` pckg. This of course does not make any harm here, but I am always a bit sceptical about using the same names as well known function names from base packages. For `f` I wanted to be consistent with `rapply` - which was an inspiration for `h_bapply`
2. Actually b can be anything, as `h_bapply` is recursive body apply. So with the first iteration it is just a body of the function but finally can be integer, character or whatsoever.... 
What do you think Daniel?

Reply:
thx, re 1: true, but can always use `base::body` if needed for the function. re 2: or just call it `x` if it can be anything?
can also consider calling it `h_rapply` if the difference is small compared to `rapply` and the point is more the recursive application rather than `body`